File: experimental/mlf-type-system/src/mlf/infer.ts
import {BindingMap} from '../utils/bindings';

import {Diagnostics} from './diagnostics';
import {Expression} from './expression';
import {Prefix} from './prefix';
import {Polytype, Type} from './type';
import {UnifyError, unify} from './unify';

export type InferError<T> =
  | UnifyError<T>
  | {
      readonly kind: 'UnboundVariable';
      readonly identifier: string;
    };

/**
 * Infers the type of an untyped expression. Detects type incompatibilities and
 * converts malformed expressions into error expressions so they will crash
 * at runtime.
 */
export function infer<Diagnostic>(
  diagnostics: Diagnostics<InferError<Diagnostic>>,
  variables: Iterable<[string, Type]>,
  expression: Expression<Diagnostic>
): Expression<InferError<Diagnostic>, Type> {
  const scope = new BindingMap(variables);
  const prefix = new Prefix();
  const result = inferExpression(diagnostics, scope, prefix, expression);
  if (!prefix.isEmpty()) {
    throw new Error('Not all type variables were cleaned up from the prefix.');
  }
  return result;
}

function inferExpression<Diagnostic>(
  diagnostics: Diagnostics<InferError<Diagnostic>>,
  scope: BindingMap<string, Type>,
  prefix: Prefix,
  expression: Expression<Diagnostic>
): Expression<InferError<Diagnostic>, Type> {
  switch (expression.description.kind) {
    // A variable references some value in the current context. The
    // variable’s type is that value’s type. If no such value exists then we
    // report an unbound variable.
    case 'Variable': {
      const variable = expression.description;
      const identifier = variable.name;
      const type = scope.get(identifier);
      if (type !== undefined) {
        return Expression.Typed.variable(type, identifier);
      } else {
        // Handle the error case by calling `inferExpression()` again so we
        // don’t duplicate logic.
        return inferExpression(
          diagnostics,
          scope,
          prefix,
          Expression.error(
            diagnostics.report({
              kind: 'UnboundVariable',
              identifier,
            })
          )
        );
      }
    }

    // Constant expressions have a native type.
    case 'Constant': {
      const constant = expression.description.constant;
      switch (constant.kind) {
        case 'Boolean':
          return Expression.Typed.boolean(constant.value);
        case 'Number':
          return Expression.Typed.number(constant.value);
        case 'String':
          return Expression.Typed.string(constant.value);
        default:
          const never: never = constant;
          return never;
      }
    }

    case 'Function': {
      const function_ = expression.description;

      // Increment the level. This is because we will generalize all dead type
      // variables at the end of type inference.
      prefix.incrementLevel();

      // Introduce a new type variable for the function parameter. Through the
      // inference of our function body we should solve this to a proper type.
      const parameterType = prefix.fresh();

      // Infer our function body type. Introducing the variable we just defined
      // into scope.
      scope.push(function_.parameter, parameterType);
      const body = inferExpression(diagnostics, scope, prefix, function_.body);
      scope.pop();

      // If the type of our body is polymorphic then we need to quantify the
      // type of our function by our body type.
      const bodyType = Type.isMonotype(body.type)
        ? body.type
        : prefix.freshWithBound(Type.flexibleBound(body.type));

      // Generalize the function type. We must generalize before deallocating
      // all the type variables at this level.
      const type = generalize(prefix, Type.function_(parameterType, bodyType));

      // Decrement the level after generalizing.
      prefix.decrementLevel();

      return Expression.Typed.function_(type, function_.parameter, body);
    }

    case 'Call': {
      const call = expression.description;

      // Increment the level. This is because we will generalize all dead type
      // variables at the end of type inference.
      prefix.incrementLevel();

      // Infer the types for our callee and argument inside of our type
      // variable quantification.
      const callee = inferExpression(diagnostics, scope, prefix, call.callee);
      const argument = inferExpression(
        diagnostics,
        scope,
        prefix,
        call.argument
      );

      // Convert the callee to a monomorphic type. If the callee type is
      // polymorphic then we need to add a type variable to our prefix.
      const calleeType = Type.isMonotype(callee.type)
        ? callee.type
        : prefix.freshWithBound(Type.flexibleBound(callee.type));

      // Convert the argument type to a monomorphic type. If the argument type
      // is polymorphic then we need to add a type variable to our prefix.
      const argumentType = Type.isMonotype(argument.type)
        ? argument.type
        : prefix.freshWithBound(Type.flexibleBound(argument.type));

      // Create a fresh type variable for the body type. This type will be
      // solved during unification.
      const bodyType = prefix.fresh();

      // Unify the type of the callee with the function type we expect. This
      // should solve any unknown type variables.
      const error = unify(
        diagnostics,
        prefix,
        calleeType,
        Type.function_(argumentType, bodyType)
      );

      // Generalize the body type and return it. We must generalize before
      // deallocating all the type variables at this level.
      const type = generalize(prefix, bodyType);

      // Decrement the level after generalizing.
      prefix.decrementLevel();

      // If there was an error during unification then we need to return an
      // error expression which will fail at runtime instead of an
      // call expression.
      return error === undefined
        ? Expression.Typed.call(type, callee, argument)
        : Expression.Typed.error(type, error);
    }

    // A binding infers a type for its value and introduces that value into
    // scope for the body. In a typical ML type system we would perform
    // generalization at let-bindings. However, in our MLF type system we keep
    // variables as polymorphic until they are applied. At which point we
    // instantiate them.
    case 'Binding': {
      // Infer bindings in a loop to avoid stack overflows from lots of
      // recursive function calls.
      const values = [];
      while (expression.description.kind === 'Binding') {
        const binding = expression.description;
        const value = inferExpression(
          diagnostics,
          scope,
          prefix,
          binding.value
        );
        scope.push(binding.name, value.type);
        values.push({name: binding.name, value});
        expression = binding.body;
      }

      // Infer the type of the body expression.
      const body = inferExpression(diagnostics, scope, prefix, expression);

      // Remove all bindings from scope and rebuild a typed binding expression.
      // Note that our bindings will be popped in the reverse order to which
      // they were pushed.
      let result = body;
      while (values.length !== 0) {
        scope.pop();
        const {name, value} = values.pop()!; // tslint:disable-line no-non-null-assertion
        result = Expression.Typed.binding(name, value, result);
      }

      return result;
    }

    // An annotation allows us to expect a type for a given expression. We
    // require type annotations for a function argument to be used
    // polymorphically. For example the auto function needs an annotation for
    // the function parameter: λx.x x
    case 'Annotation': {
      const annotation = expression.description;

      // Increment the level before creating new type variables.
      prefix.incrementLevel();

      // Infer the annotation expression.
      const value = inferExpression(
        diagnostics,
        scope,
        prefix,
        annotation.value
      );

      // Produce a monotype for the value type.
      const valueType = Type.isMonotype(value.type)
        ? value.type
        : prefix.freshWithBound(Type.flexibleBound(value.type));

      // Produce a monotype for the annotation type.
      const annotationType = Type.isMonotype(annotation.type)
        ? annotation.type
        : prefix.freshWithBound(Type.rigidBound(annotation.type));

      // Unify the value and annotation types.
      const error = unify(diagnostics, prefix, valueType, annotationType);

      // Decrement the level and destroy the type variables we created.
      prefix.decrementLevel();

      // If unification was not a success then we return an error expression.
      // The error expression is still of the annotation type, however.
      return error === undefined
        ? Expression.Typed.annotation(value, annotation.type)
        : Expression.Typed.error(annotation.type, error);
    }

    // Runtime errors have the bottom type since they will crash at runtime.
    case 'Error':
      return Expression.Typed.error(Type.bottom, expression.description.error);

    default:
      const never: never = expression.description;
      return never;
  }
}

/**
 * Turns type variables at a level larger then the current level into generic
 * quantified type bounds.
 */
function generalize(prefix: Prefix, type: Polytype): Polytype {
  const quantify = new Set<string>();
  generalize(quantify, prefix, type);
  // Quantify our type for every variable in the `quantify` set. The order of
  // the `quantify` set does matter! Since some type variables may have a
  // dependency on others.
  const quantifyReverse = Array(quantify.size);
  let i = 0;
  for (const name of quantify) {
    quantifyReverse[quantify.size - i++ - 1] = name;
  }
  for (const name of quantifyReverse) {
    const {bound} = prefix.lookup(name);
    type = Type.quantify(name, bound, type);
  }
  return type;

  // Iterate over every free type variable and determine if we need to
  // quantify it.
  function generalize(quantify: Set<string>, prefix: Prefix, type: Polytype) {
    for (const name of Type.getFreeVariables(type)) {
      const {level, bound} = prefix.lookup(name);
      // If we have a type variable with a level greater than our current
      // level then we need to quantify that type variable.
      if (level >= prefix.getLevel()) {
        // If `quantify` already contains this type variable then we don’t
        // need to add it again.
        if (!quantify.has(name)) {
          // Generalize the dead type variables in our bound as well.
          generalize(quantify, prefix, bound.type);
          // It is important that we add the name to `quantify` _after_ we
          // generalize the bound type. The order of variables in `quantify`
          // does matter.
          quantify.add(name);
        }
      }
    }
  }
}

File: src/Infer.ml
(* IMPORTANT: It is expected that all free types in context are bound in the
 * prefix! If this is not true then we may panic.
 *
 * Type inference looks at an expression and produces a principal type for the
 * expression. The core algorithm is described in section 7.1 of the
 * [MLF Thesis][1].
 *
 * In theory:
 *
 * > A type inference problem is a triple `(Q, Γ, a)`, where all free type
 * > variables in `Γ` are bound in `Q`. A pair `(Q', o)` is a solution to this
 * > problem if `Q ⊑ Q'` and `(Q') Γ ⊦ a : o` holds.
 *
 * [1]: https://pastel.archives-ouvertes.fr/file/index/docid/47191/filename/tel-00007132.pdf *)
let rec infer prefix context expression =
  match expression.Expression.description with
  (* If the variable exists in context then return its type. Otherwise report an
   * unbound variable error and return the bottom type since the code will panic
   * at runtime. *)
  | Variable { name } -> (
    match StringMap.find_opt name context with
    | Some t -> t
    | None ->
      let _ = Diagnostic.report_error (UnboundVariable { name }) in
      Type.bottom
  )

  (* Return the constant types for constants. *)
  | Boolean _ -> Type.to_polytype Type.boolean
  | Number _ -> Type.to_polytype Type.number
  | String _ -> Type.to_polytype Type.string

  (* Functions create a fresh type variable for their parameters and then infer
   * the body type with that type variable. It is expected that through
   * unification the parameter type will be solved. At the very end we
   * generalize the type variables created at this level which weren’t updated
   * to a higher level. *)
  | Function { parameter; body } ->
    Prefix.level prefix (fun () -> (
      let parameter_type = Prefix.fresh prefix in
      let context = StringMap.add parameter (Type.to_polytype parameter_type) context in
      let body_type = infer prefix context body in
      let body_type = Prefix.fresh_with_bound prefix (Type.bound Flexible body_type) in
      Prefix.generalize prefix (Type.function_ parameter_type body_type)
    ))

  (* For function calls infer the type of the callee and argument. Then unify
   * the type of the callee with a function type with the argument as the
   * parameter and a fresh type variable as the body. Through unification the
   * fresh type variable should be solved for. At the very end generalize the
   * body type and associated referenced type variables created at this level
   * which weren’t updated to a higher level. *)
  | Call { callee; argument } ->
    let callee_type = infer prefix context callee in
    let argument_type = infer prefix context argument in
    Prefix.level prefix (fun () -> (
      let callee_type = Prefix.fresh_with_bound prefix (Type.bound Flexible callee_type) in
      let argument_type = Prefix.fresh_with_bound prefix (Type.bound Flexible argument_type) in
      let body_type = Prefix.fresh prefix in
      let _ = Unify.unify prefix callee_type (Type.function_ argument_type body_type) in
      Prefix.generalize prefix body_type
    ))

  (* With bindings we infer the value for the type, add it to our context under
   * the provided name, and then infer the type for our body. *)
  | Binding { name; value; body } ->
    let value_type = infer prefix context value in
    let context = StringMap.add name value_type context in
    infer prefix context body

  (* A type annotation tests whether the annotated value has a type equivalent
   * to the annotation. We test the equivalence of the two types with
   * unification.
   *
   * Notably we use a rigid type bound for the annotation. This is because we
   * don’t want to instantiate bounds of the annotation. Otherwise we might let
   * `(add1: ∀a.a → a)` pass type checking! This would allow us to pass _any
   * value_ to the `add1` function whose signature is `number → number`. This
   * would be bad. We need to error. *)
  | Annotation { value; type_ } ->
    (* TODO: Make sure that `type_` does not have unbound type variables! *)
    let value_type = infer prefix context value in
    Prefix.level prefix (fun () -> (
      let value_type = Prefix.fresh_with_bound prefix (Type.bound Flexible value_type) in
      let type_' = Prefix.fresh_with_bound prefix (Type.bound Rigid type_) in
      let _ = Unify.unify prefix value_type type_' in
      type_
    ))
